stan/math/opencl/add.hpp

```diff
@@ -27,10 +27,10 @@ inline matrix_cl add(const matrix_cl& A, const matrix_cl& B) {
   if (C.size() == 0) {
     return C;
   }
-  cl::CommandQueue cmdQueue = opencl_context.queue();
   try {
-    opencl_kernels::add(cl::NDRange(A.rows(), A.cols()), C.buffer(), A.buffer(),
-                        B.buffer(), A.rows(), A.cols());
+    cl::Event add_event = opencl_kernels::add(cl::NDRange(A.rows(), A.cols()),
```

Comment:
> First off, this is all awesome and I love how the interface is evolving. And the speedups are very impressive. 

Ty and same!

> I was wondering if we can have the global_range_kernel et al add the Event to the matrix themselves? this might require going to a convention where the output of the kernel always comes first, for example.

We could do that, but personally I'd prefer that the developer explicitly says, "This is the matrix we have to wait for". idt we do it anywhere today, but I could imagine an instance of us having two output two matrices in one kernel. If we wanted to remove that explicit line we could template the kernel call to look something like.

```cpp
opencl_kernels::add<First, Second>(cl::NDRange(A.rows(), A.cols()), ...);
```
> Where First and Second are enums and tell us which buffers we need to attach events to. I think that's possible? More of a nicety than anything. Could wait for another PR

> Separately it might be cool to just have the kernels always pass the matrix_cl sizes in for all their arguments in order, and then to let the kernel code itself discard the ones it doesn't need. Then we get rid of all the A.rows() etc even though we're passing in A.

I'm not totally sure how to go about this in a clean way. I'm worried that would start obscuring the connection between what's called in the kernel and the top level interface a bit too much

> Though we probably still need the NDRanges in the signature here, I remember that was basically impossible to get rid of.

The NDRange is needed.

One extra thing here. I think with the new templating we can actually get rid of the `global` and `local` kernel structs and just have one. Then the template Args would just take in either one or two NDRanges. We can defer users in the docs to the docs for the OpenCL functor this calls to see what is available.

Reply:
ah that does seem right, very cool. We can leave the investigation of passing of sizes and global/local for another issue/PR. I think for this PR we should just figure out if we want to expose the events to the outside world outside of the global/local_range_kernels (and rename events(e) to add_event(e)).